google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java

```diff
@@ -95,6 +95,186 @@ public BucketInfo apply(com.google.api.services.storage.model.Bucket pb) {
+        return new IamConfiguration(this);
+      }
+    }
+  }
+
+  /**
+   * Configuration of BucketPolicyOnly for a bucket. When this is configured to enable BucketPolicyOnly and included
+   * in an {@code IamConfiguration}, legacy ACL access to the bucket will be disabled, and all authorization policies
+   * will be configured through IAM.
+   *
+   * @see <a href="https://cloud.google.com/storage/docs/bucket-policy-only">Bucket Policy Only</a>
+   */
+  public static class BucketPolicyOnly implements Serializable {
+    private static final long serialVersionUID = 2260445211318576550L;
+    private final Boolean enabled;
+    private final DateTime lockedTime;
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      BucketPolicyOnly other = (BucketPolicyOnly) o;
+      return Objects.equals(toPb(), other.toPb());
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(enabled, lockedTime);
+    }
+
+    public Boolean getEnabled() {
+      return enabled;
+    }
+
+    public DateTime getLockedTime() {
+      return lockedTime;
+    }
+
+    public static Builder newBuilder() {
+      return new Builder();
+    }
+
+    Bucket.IamConfiguration.BucketPolicyOnly toPb() {
+      Bucket.IamConfiguration.BucketPolicyOnly bucketPolicyOnly = new Bucket.IamConfiguration.BucketPolicyOnly();
+
+      bucketPolicyOnly.setEnabled(enabled);
+      bucketPolicyOnly.setLockedTime(lockedTime);
+
+      return bucketPolicyOnly;
+    }
+
+    static BucketPolicyOnly fromPb(Bucket.IamConfiguration.BucketPolicyOnly bucketPolicyOnly) {
+      return newBuilder()
+              .setEnabled(bucketPolicyOnly.getEnabled())
+              .setLockedTime(bucketPolicyOnly.getLockedTime())
+              .build();
+    }
+
+    private BucketPolicyOnly(Builder builder) {
+      this.enabled = builder.enabled;
+      this.lockedTime = builder.lockedTime;
+    }
+
+    public Builder toBuilder() {
+      Builder builder = new Builder();
+      builder.setLockedTime(lockedTime);
+      builder.setEnabled(enabled);
+      return builder;
+    }
+
+    /** Builder for {@code BucketPolicyOnly} */
+    public static class Builder {
+      private Boolean enabled;
+      private DateTime lockedTime;
+
+      private Builder() {}
+
+      /** Sets whether BucketPolicyOnly is enabled for this bucket. When this is enabled, access to the bucket will be
+       * configured through IAM, and legacy ACL policies will not work. All legacy ACL policies must be removed from
+       * the bucket's info before this can be enabled. When this is first enabled, {@code lockedTime} will be set by
+       * the API automatically. This field can then be disabled until the time specified, after which it will
+       * become immutable and calls to change it will fail. If this is enabled, calls to access legacy ACL information
+       * will fail.
```

Comment:
The following is no longer true:
> All legacy ACL policies must be removed from the bucket's info before this can be enabled.

Are you experiencing something different?  If you're referring to the following:
```java
if (bucket.getIamConfiguration() != null && bucket.getIamConfiguration().getBucketPolicyOnly() != null
              && bucket.getIamConfiguration().getBucketPolicyOnly().getEnabled()) {
     //If BucketPolicyOnly is enabled, patch calls will fail if ACL information is included in the request
     bucket.setDefaultObjectAcl(null);
     bucket.setAcl(null);

    if (projection == null) {
       projection = NO_ACL_PROJECTION;
    }
}
```

`setAcl(null)` and `setDefaultObjectAcl(null)` is required if these fields are set in the metadata resource when performing a patch. If they weren't set then this would succeed. It's more of a client case.